test/unit/rfc/paths_rule.cpp

```diff
@@ -190,13 +192,68 @@ class paths_rule_test
         bad<T>("/%25%2e");
     }
 
+    void
+    test_normalize()
+    {
+        {
+            auto check = [](
+                   string_view p,
+                   string_view e1,
+                   string_view e2) {
+                std::allocator<char> a{};
+                auto s1 = detail::remove_dot_segments(p, a, true);
```

Comment:
You should avoid invoking non-public functions in tests, when possible. In other words it is better to test using what users will actually be calling.

Reply:
I considered that and it's what I did for other tests. The problem is this algorithm is quite complex and there are lots of sensitive code paths and conditions to test. We could move that to `url_view.cpp` and find some URLs that trigger the same code paths, but the tests are going to become a little bulkier.

Reply:
You can't move it to url_view.cpp you have to put it in url.cpp, because it has to modify the segments no? But yes, it is better to get full coverage on `remove_dot_segments` using public APIs instead of invoking the implementation function directly. It will require more tests but it is worth it.